k9mail/src/main/java/com/fsck/k9/planck/ui/activities/provisioning/ProvisioningActivity.kt

```diff
@@ -1,53 +1,119 @@
+    private lateinit var folderPickerLauncher: ActivityResultLauncher<Intent>
+    private val viewModel: ProvisioningViewModel by viewModels()
     private lateinit var waitingForProvisioningText: TextView
     private lateinit var progressBar: ProgressBar
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setupViews()
+        observeViewModel()
     }
 
-    private fun setupViews() {
-        setContentView(R.layout.activity_provisioning)
-        waitingForProvisioningText = findViewById(R.id.waitingForProvisioningText)
-        progressBar = findViewById(R.id.provisioning_progress)
+    private fun observeViewModel() {
+        this.viewModel.state.observe(this) { state ->
+            state?.let { renderState(state) }
+        }
     }
 
-    override fun onStart() {
-        super.onStart()
-        presenter.attach(this)
+    private fun renderState(state: ProvisionState) {
+        when (state) {
+            is ProvisionState.WaitingForProvisioning ->
+
+                waitingForProvisioning()
+
+            is ProvisionState.InProvisioning ->
+                provisioningProgress()
+
+            is ProvisionState.WaitingToInitialize -> {
+                if (state.offerRestore) {
+                    offerRestorePlanckData()
+                } else {
+                    initializing()
+                }
+            }
+
+            is ProvisionState.Initializing ->
+                if (state.provisioned) {
+                    initializingAfterSuccessfulProvision()
+                } else {
+                    initializing()
+                }
+
+            is ProvisionState.Initialized ->
+                initialized()
+
+            is ProvisionState.Error -> {
+                val throwableMessage = state.throwable.message
+                val message =
+                    if (throwableMessage.isNullOrBlank())
+                        state.throwable.stackTraceToString()
+                    else throwableMessage
+                if (state.throwable is ProvisioningFailedException) {
+                    displayProvisioningError(message)
+                } else {
+                    displayInitializationError(message)
+                }
+            }
+        }
+    }
+
+    private fun setupViews() {
+        binding = ActivityProvisioningBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+        waitingForProvisioningText = binding.waitingForProvisioningText
+        progressBar = binding.provisioningProgress
+        binding.provisioningSkipButton.setOnClickListener { this.viewModel.initializeApp() }
+        binding.provisioningRestoreDataButton.setOnClickListener { pickFolder() }
+
+        folderPickerLauncher =
+            registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
+                Log.e("EFA-625", "RESULT CODE: ${result.resultCode}, DATA: ${result.data}")
```

Comment:
so we want to reference the ticket ID in the log entry? What's the upside?

Reply:
it's just a debug line I guess

Reply:
was the moment of change.

Reply:
removed it thanks.